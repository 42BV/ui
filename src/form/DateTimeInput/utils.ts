import moment from 'moment';
import { InputMask } from '../Input/Input';
import { DateFormat, TimeFormat } from './types';

export function combineFormat(
  dateFormat: DateFormat,
  timeFormat: TimeFormat
): string {
  if (dateFormat !== false && timeFormat !== false) {
    return `${dateFormat} ${timeFormat}`;
  } else if (dateFormat !== false) {
    return dateFormat;
  } else if (timeFormat !== false) {
    return timeFormat;
  }

  throw new Error(
    'DateTimeInput: dateFormat and timeFormat cannot both be false'
  );
}

export function formatToMask(
  dateFormat: DateFormat,
  timeFormat: TimeFormat
): InputMask {
  const dateMask = dateFormatToMask(dateFormat);

  const timeMask = timeFormatToMask(timeFormat);

  const dateMaskIsDefined = dateMask.length > 0;

  if (dateMaskIsDefined && timeMask.length > 0) {
    return [...dateMask, ' ', ...timeMask];
  } else if (dateMaskIsDefined) {
    return dateMask;
  } else {
    return timeMask;
  }
}

function timeFormatToMask(timeFormat: TimeFormat): InputMask {
  if (timeFormat === false) {
    return [];
  }

  return timeFormat.split('').map((char) => (char === ':' ? ':' : /\d/));
}

function dateFormatToMask(dateFormat: DateFormat): InputMask {
  if (dateFormat === false) {
    return [];
  }

  const separator = extractSeparator(dateFormat);
  return dateFormat
    .split('')
    .map((char) => (char === separator ? separator : /\d/));
}

function extractSeparator(dateFormat: string): string {
  if (dateFormat.includes('-')) {
    return '-';
  }

  if (dateFormat.includes('/')) {
    return '/';
  }

  if (dateFormat.includes('.')) {
    return '.';
  }

  throw new Error('DateTimeInput: cannot extract separator from dateFormat');
}

export function isDate(
  value: string,
  dateFormat: DateFormat,
  timeFormat: TimeFormat
): boolean {
  // _ indicates a masks
  if (value.includes('_')) {
    return false;
  } else {
    const date = moment(
      value.trim(), // value includes an empty char at the back for some reason.
      combineFormat(dateFormat, timeFormat)
    );

    return date.isValid();
  }
}
