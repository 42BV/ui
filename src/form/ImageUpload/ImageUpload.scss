.img-upload {
  position: relative;

  &:hover {
    .img-upload-wrapper {
      background: var(--bs-gray-300);
      cursor: pointer;
    }
  }

  input[type='file'] {
    opacity: 0;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 100%;

    &:hover {
      cursor: pointer;
    }
  }

  canvas {
    max-width: 100%;
    height: auto !important;
  }

  .img-upload-wrapper {
    border: 0.625rem dashed var(--bs-gray-400);
    padding-top: 3rem;
    padding-bottom: 3rem;
    text-align: center;

    > i {
      font-size: 5rem;
    }
  }
}
