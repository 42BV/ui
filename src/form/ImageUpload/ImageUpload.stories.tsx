import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import ImageUpload, { JarbImageUpload, requireImage } from './ImageUpload';
import { Form, FinalForm } from '../story-utils';
import { Tooltip, Icon } from '../..';
import FileInput from '../FileInput/FileInput';

storiesOf('Form/ImageUpload', module)
  .add('rect', () => {
    return (
      <Form>
        <ImageUpload
          id="image-uploader"
          label="Profile photo"
          crop={{
            type: 'rect',
            width: 500,
            height: 400
          }}
          onChange={(value) => action(`You entered ${value}`)}
        />
      </Form>
    );
  })
  .add('circle', () => {
    return (
      <Form>
        <ImageUpload
          id="image-uploader"
          label="Profile photo"
          crop={{
            type: 'circle',
            size: 250
          }}
          onChange={(value) => action(`You entered ${value}`)}
        />
      </Form>
    );
  })
  .add('without label', () => {
    return (
      <Form>
        <ImageUpload
          id="image-uploader"
          crop={{
            type: 'rect',
            width: 500,
            height: 400
          }}
          onChange={(value) => action(`You entered ${value}`)}
        />
      </Form>
    );
  })
  .add('with custom label', () => {
    return (
      <Form>
        <ImageUpload
          id="image-uploader"
          label={
            <div className="d-flex justify-content-between">
              <span>Profile photo</span>
              <Tooltip
                className="ml-1"
                style={{ zIndex: 101 }}
                content="You can edit the photo after you've selected it"
              >
                <Icon icon="info" />
              </Tooltip>
            </div>
          }
          crop={{
            type: 'rect',
            width: 500,
            height: 400
          }}
          onChange={(value) => action(`You entered ${value}`)}
        />
      </Form>
    );
  })
  .add('File as initial value', () => {
    const [file, setFile] = useState<File | null>();
    return (
      <Form>
        <FileInput
          id="image-uploader"
          label="Choose an image"
          accept="image/*"
          onChange={setFile}
        />
        {file ? (
          <ImageUpload
            id="image-uploader"
            label="Profile photo"
            crop={{
              type: 'rect',
              width: 500,
              height: 400
            }}
            value={file}
            onChange={setFile}
          />
        ) : null}
      </Form>
    );
  })
  .add('jarb', () => {
    return (
      <FinalForm>
        <JarbImageUpload
          id="image-uploader"
          name="profile"
          label="Profile photo"
          validators={[requireImage('Profile photo')]}
          crop={{
            type: 'rect',
            width: 500,
            height: 400
          }}
          jarb={{
            validator: 'User.profile',
            label: 'Profile photo'
          }}
        />
      </FinalForm>
    );
  });
