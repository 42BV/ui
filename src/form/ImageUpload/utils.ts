import picaFn from 'pica';
import { ImageUploadCrop } from './ImageUpload';
import { Buffer } from 'buffer';
Buffer.from('anything','base64');

let picaInstance: any;

export function getPicaInstance() {
  if (picaInstance) {
    return picaInstance;
  }

  picaInstance = picaFn();

  return picaInstance;
}

export function dataUrlToFile(dataUrl: string, fileName: string): File {
  const byteString = Buffer.from(dataUrl.split(',')[1], 'base64');

  // separate out the mime component
  const mimeString = dataUrl.split(',')[0].split(':')[1].split(';')[0];

  // write the bytes of the string to a typed array
  const bytes = new Uint8Array(byteString.length);
  for (let i = 0; i < byteString.length; i++) {
    bytes[i] = byteString[i];
  }

  return new File([ bytes ], fileName, { type: mimeString });
}

type AvatarEditorConfig = {
  borderRadius: number;
  width: number;
  height: number;
};

export function cropToAvatarEditorConfig(
  crop: ImageUploadCrop
): AvatarEditorConfig {
  if (crop.type === 'rect') {
    const { width, height } = crop;

    return { borderRadius: 0, width, height };
  } else {
    const { size } = crop;

    /* 
      When Crop is 'circle' a ridiculously high border radius is needed 
      so crop.size of different sizes show a perfect circle. Not sure
      why this occurs.
    */
    const borderRadius = 9000000;

    return { borderRadius, width: size, height: size };
  }
}

// Calculate the next 'scale' to use based on how hard the user scrolled
export function calculateScale(scale: number, delta: number): number {
  const zoomFactor = Math.max(Math.abs(delta) / 1000, 0.1);
  const nextScale = delta > 0 ? scale - zoomFactor : scale + zoomFactor;
  return Math.min(10, Math.max(1, nextScale));
}

// Replace any part after the last dot in the filename with .png
export function replaceFileExtension(fileName: string): string {
  return fileName.replace(/\.\w+$/, '.png');
}

export function readFile(file: File, callback: (result: string) => void) {
  const reader = new FileReader();
  reader.onloadend = () => {
    /* istanbul ignore else */
    if (typeof reader.result === 'string') {
      callback(reader.result);
    }
  };
  reader.readAsDataURL(file);
}
