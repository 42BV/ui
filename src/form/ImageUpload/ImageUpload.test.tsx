import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as imgUploadUtils from './utils';

import ImageUpload, {
  ImageUploadCrop,
  limitImageSize,
  requireImage,
  Text
} from './ImageUpload';

import * as testUtils from '../../test/utils';

describe('Component: ImageUpload', () => {
  let imgUpload: ShallowWrapper;

  let onChangeSpy: jest.Mock;
  let onBlurSpy: jest.Mock;

  function setup({
    value,
    cropType,
    text,
    hasLabel = true,
    emptyLabel = false,
    keepOriginalFileExtension
  }: {
    value?: File | string;
    cropType: 'rect' | 'circle';
    text?: Text;
    hasLabel?: boolean;
    emptyLabel?: boolean;
    keepOriginalFileExtension?: boolean;
  }) {
    onChangeSpy = jest.fn();
    onBlurSpy = jest.fn();

    const crop: ImageUploadCrop =
      cropType === 'rect'
        ? { width: 500, height: 500, type: 'rect' }
        : { size: 250, type: 'circle' };

    const props = {
      crop,
      value,
      onChange: onChangeSpy,
      onBlur: onBlurSpy,
      error: 'Some error',
      text,
      keepOriginalFileExtension
    };

    if (hasLabel) {
      imgUpload = shallow(
        <ImageUpload
          id="image-uploader"
          label={emptyLabel ? '' : 'Profile photo'}
          color="success"
          {...props}
        />
      );
    } else {
      imgUpload = shallow(<ImageUpload color="success" {...props} />);
    }
  }

  describe('componentDidMount', () => {
    it('should not show the image when the value is null', () => {
      // @ts-expect-error Test mock
      const imgUpload = new ImageUpload();

      // @ts-expect-error Test mock
      imgUpload.props = { value: null, id: 1 };
      jest.spyOn(imgUpload, 'setState');

      imgUpload.componentDidMount();

      expect(imgUpload.setState).toHaveBeenCalledTimes(0);
    });

    it('should not show the image when the value is empty string', () => {
      // @ts-expect-error Test mock
      const imgUpload = new ImageUpload();

      // @ts-expect-error Test mock
      imgUpload.props = { value: '' };
      jest.spyOn(imgUpload, 'setState');

      imgUpload.componentDidMount();

      expect(imgUpload.setState).toHaveBeenCalledTimes(0);
    });

    test('it should show the image when the value is a non empty string', () => {
      // @ts-expect-error Test mock
      const imgUpload = new ImageUpload();

      // @ts-expect-error Test mock
      imgUpload.props = { value: 'maarten.png' };
      jest.spyOn(imgUpload, 'setState').mockImplementation(() => undefined);

      imgUpload.componentDidMount();

      expect(imgUpload.setState).toHaveBeenCalledTimes(1);
      expect(imgUpload.setState).toHaveBeenCalledWith({
        mode: 'file-selected',
        imageSrc: 'maarten.png'
      });
    });

    test('it should show the image when the value is a File', () => {
      // @ts-expect-error Test mock
      const imgUpload = new ImageUpload();

      // @ts-expect-error Test mock
      imgUpload.props = { value: new File([''], 'maarten.png') };
      jest.spyOn(imgUpload, 'setState').mockImplementation(() => undefined);
      jest
        .spyOn(imgUpload, 'readFile')
        .mockImplementation((file: File, callback: (result: string) => void) =>
          callback('test')
        );

      imgUpload.componentDidMount();

      expect(imgUpload.setState).toHaveBeenCalledTimes(1);
      expect(imgUpload.setState).toHaveBeenCalledWith({
        mode: 'file-selected',
        fileName: 'maarten.png',
        imageSrc: 'test'
      });
    });
  });

  describe('ui', () => {
    test('file-selected as rect', () => {
      setup({
        value: new File([''], 'maarten.png'),
        cropType: 'rect'
      });

      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'file-selected' });

      expect(toJson(imgUpload)).toMatchSnapshot(
        'Component: ImageUpload => ui => file-selected as rect'
      );
    });

    test('file-selected as circle', () => {
      setup({
        value: new File([''], 'maarten.png'),
        cropType: 'circle',
        text: {
          cancel: 'CANCEL',
          change: 'CHANGE',
          remove: 'REMOVE',
          done: 'DONE'
        }
      });

      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'file-selected' });

      expect(toJson(imgUpload)).toMatchSnapshot(
        'Component: ImageUpload => ui => file-selected as circle'
      );
    });

    test('edit as rect', () => {
      setup({
        value: undefined,
        cropType: 'rect',
        text: {
          cancel: 'CANCEL',
          change: 'CHANGE',
          remove: 'REMOVE',
          done: 'DONE'
        }
      });

      imgUpload.setState({
        imageSrc: 'maarten.png',
        mode: 'edit'
      });

      expect(toJson(imgUpload)).toMatchSnapshot(
        'Component: ImageUpload => ui => edit as rect'
      );
    });

    test('edit as circle', () => {
      setup({ value: undefined, cropType: 'circle' });

      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'edit' });

      expect(toJson(imgUpload)).toMatchSnapshot(
        'Component: ImageUpload => ui => edit as circle'
      );
    });

    test('no-file', () => {
      setup({ value: 'maarten.png', cropType: 'rect' });
      imgUpload.setState({ mode: 'no-file' });

      expect(toJson(imgUpload)).toMatchSnapshot(
        'Component: ImageUpload => ui => no-file'
      );
    });

    test('without label', () => {
      setup({ cropType: 'rect', hasLabel: false });

      expect(toJson(imgUpload)).toMatchSnapshot(
        'Component: ImageUpload => ui => without label'
      );
    });

    test('with empty label', () => {
      setup({ cropType: 'rect', emptyLabel: true });

      expect(toJson(imgUpload)).toMatchSnapshot(
        'Component: ImageUpload => ui => with empty label'
      );
    });

    test('file-selected as rect without label', () => {
      setup({
        value: new File([''], 'maarten.png'),
        cropType: 'rect',
        hasLabel: false
      });

      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'file-selected' });

      expect(toJson(imgUpload)).toMatchSnapshot(
        'Component: ImageUpload => ui => file-selected as rect without label'
      );
    });
  });

  describe('mode: no-file events', () => {
    it('should when file is selected transition to edit mode', (done) => {
      setup({ value: undefined, cropType: 'circle' });

      const instance = imgUpload.instance();
      jest.spyOn(instance, 'setState');

      const file = new File([''], 'maarten.png');

      // @ts-expect-error Test mock
      imgUpload
        .find('input')
        .props()
        // @ts-expect-error Test mock
        .onChange({ target: { files: [file] } });

      // Wait for reader to have finished reading the file
      testUtils.waitForUI(() => {
        expect(instance.setState).toHaveBeenCalledTimes(1);
        expect(instance.setState).toHaveBeenCalledWith({
          imageSrc: 'data:application/octet-stream;base64,',
          mode: 'edit',
          fileName: 'maarten.png',
          rotate: 0,
          scale: 1
        });
        done();
      }, 1000);
    });
  });

  describe('mode: edit events', () => {
    it('should rotate left when rotate left rotation button is clicked', () => {
      setup({ value: undefined, cropType: 'circle' });
      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'edit' });

      const instance = imgUpload.instance();
      jest.spyOn(instance, 'setState');

      imgUpload.find('Button').at(0).simulate('click');

      expect(instance.setState).toHaveBeenCalledTimes(1);
      expect(instance.setState).toHaveBeenCalledWith({
        rotate: -90
      });
    });

    it('should rotate right when rotate right rotation button is clicked', () => {
      setup({ value: undefined, cropType: 'circle' });
      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'edit' });

      const instance = imgUpload.instance();
      jest.spyOn(instance, 'setState');

      imgUpload.find('Button').at(1).simulate('click');

      expect(instance.setState).toHaveBeenCalledTimes(1);
      expect(instance.setState).toHaveBeenCalledWith({
        rotate: 90
      });
    });

    it('should scale the image when mouse is scrolled on avatar editor', () => {
      setup({ value: undefined, cropType: 'circle' });
      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'edit', scale: 20 });

      jest.spyOn(imgUploadUtils, 'calculateScale').mockReturnValue(42);

      const instance = imgUpload.instance();
      jest.spyOn(instance, 'setState');

      const preventDefault = jest.fn();

      // @ts-expect-error Test mock
      imgUpload
        .find('div')
        .at(2)
        .props()
        // @ts-expect-error Test mock
        .onWheel({ deltaY: 10, preventDefault });

      expect(preventDefault).toHaveBeenCalledTimes(1);

      expect(imgUploadUtils.calculateScale).toHaveBeenCalledTimes(1);
      expect(imgUploadUtils.calculateScale).toHaveBeenCalledWith(20, 10);

      expect(instance.setState).toHaveBeenCalledTimes(1);
      expect(instance.setState).toHaveBeenCalledWith({
        scale: 42
      });
    });

    it('should reset the file input when cancel button is clicked and move to state no-file', () => {
      setup({ value: undefined, cropType: 'circle' });
      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'edit' });

      const instance = imgUpload.instance();
      // @ts-expect-error Test mock
      instance.inputRef = { current: null };

      imgUpload.find('Button').at(2).simulate('click');

      expect(imgUpload.state()).toEqual({
        fileName: '',
        imageSrc: '',
        mode: 'no-file',
        rotate: 0,
        scale: 1
      });

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(null);

      expect(onBlurSpy).toHaveBeenCalledTimes(1);
    });

    it('should set the final image when done button is clicked and move to state file-selected', async (done) => {
      expect.assertions(10);

      setup({ value: undefined, cropType: 'circle' });

      const { promise, resolve } = testUtils.resolvablePromise();

      const resizeSpy = jest.fn(() => {
        return promise;
      });

      jest.spyOn(imgUploadUtils, 'getPicaInstance').mockImplementation(() => {
        return {
          resize: resizeSpy
        };
      });

      const file = new File([''], 'maarten.png');

      jest
        .spyOn(imgUploadUtils, 'dataUrlToFile')
        .mockImplementation(() => file);

      imgUpload.setState({
        imageSrc: 'maarten.png',
        mode: 'edit',
        fileName: 'maarten.png'
      });

      const instance = imgUpload.instance();
      // @ts-expect-error Test mock
      instance.editorRef = {
        current: { getImage: jest.fn(() => 'fakeCanvas') }
      };
      jest.spyOn(instance, 'setState');

      try {
        await imgUpload.find('Button').at(3).simulate('click');

        resolve({
          toDataURL: jest.fn(() => 'Some base 64 string')
        });

        await promise;

        expect(resizeSpy).toHaveBeenCalledTimes(1);
        // @ts-expect-error Test mock
        expect(resizeSpy.mock.calls[0][0]).toBe('fakeCanvas');
        // @ts-expect-error Test mock
        expect(resizeSpy.mock.calls[0][2]).toEqual({ alpha: true });

        // @ts-expect-error Test mock
        const offScreenCanvas = resizeSpy.mock.calls[0][1];
        // @ts-expect-error Test mock
        expect(offScreenCanvas.width).toBe(250);
        // @ts-expect-error Test mock
        expect(offScreenCanvas.height).toBe(250);

        expect(onChangeSpy).toHaveBeenCalledTimes(1);
        expect(onChangeSpy).toHaveBeenCalledWith(file);

        expect(onBlurSpy).toHaveBeenCalledTimes(1);

        expect(instance.setState).toHaveBeenCalledTimes(1);
        expect(instance.setState).toHaveBeenCalledWith({
          mode: 'file-selected',
          imageSrc: 'Some base 64 string',
          fileName: 'maarten.png'
        });

        done();
      } catch (error) {
        console.error(error);
        done.fail();
      }
    });

    it('should replace the file extension when keepOriginalFileExtension is undefined', async (done) => {
      expect.assertions(2);

      setup({ value: undefined, cropType: 'circle' });

      const { promise, resolve } = testUtils.resolvablePromise();

      const resizeSpy = jest.fn(() => {
        return promise;
      });

      jest.spyOn(imgUploadUtils, 'getPicaInstance').mockImplementation(() => {
        return {
          resize: resizeSpy
        };
      });

      const file = new File([''], 'maarten.jpg');

      jest
        .spyOn(imgUploadUtils, 'dataUrlToFile')
        .mockImplementation(() => file);

      imgUpload.setState({
        imageSrc: 'maarten.jpg',
        mode: 'edit',
        fileName: 'maarten.jpg'
      });

      const instance = imgUpload.instance();
      // @ts-expect-error Test mock
      instance.editorRef = {
        current: { getImage: jest.fn(() => 'fakeCanvas') }
      };
      jest.spyOn(instance, 'setState');

      try {
        await imgUpload.find('Button').at(3).simulate('click');

        resolve({
          toDataURL: jest.fn(() => 'Some base 64 string')
        });

        await promise;

        expect(instance.setState).toHaveBeenCalledTimes(1);
        expect(instance.setState).toHaveBeenCalledWith({
          mode: 'file-selected',
          imageSrc: 'Some base 64 string',
          fileName: 'maarten.png'
        });

        done();
      } catch (error) {
        console.error(error);
        done.fail();
      }
    });

    it('should not replace the file extension when keepOriginalFileExtension is true', async (done) => {
      expect.assertions(2);

      setup({
        value: undefined,
        cropType: 'circle',
        keepOriginalFileExtension: true
      });

      const { promise, resolve } = testUtils.resolvablePromise();

      const resizeSpy = jest.fn(() => {
        return promise;
      });

      jest.spyOn(imgUploadUtils, 'getPicaInstance').mockImplementation(() => {
        return {
          resize: resizeSpy
        };
      });

      const file = new File([''], 'maarten.jpg');

      jest
        .spyOn(imgUploadUtils, 'dataUrlToFile')
        .mockImplementation(() => file);

      imgUpload.setState({
        imageSrc: 'maarten.jpg',
        mode: 'edit',
        fileName: 'maarten.jpg'
      });

      const instance = imgUpload.instance();
      // @ts-expect-error Test mock
      instance.editorRef = {
        current: { getImage: jest.fn(() => 'fakeCanvas') }
      };
      jest.spyOn(instance, 'setState');

      try {
        await imgUpload.find('Button').at(3).simulate('click');

        resolve({
          toDataURL: jest.fn(() => 'Some base 64 string')
        });

        await promise;

        expect(instance.setState).toHaveBeenCalledTimes(1);
        expect(instance.setState).toHaveBeenCalledWith({
          mode: 'file-selected',
          imageSrc: 'Some base 64 string',
          fileName: 'maarten.jpg'
        });

        done();
      } catch (error) {
        console.error(error);
        done.fail();
      }
    });
  });

  describe('mode: file-selected events', () => {
    it('should reset to no-file when clicking the remove button ', () => {
      setup({ value: new File([''], 'maarten.png'), cropType: 'rect' });
      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'file-selected' });

      // @ts-expect-error Test mock
      imgUpload.instance().inputRef = { current: { value: 'maarten.png' } };

      // @ts-expect-error Test mock
      imgUpload.find('Button').at(1).prop('onClick')();

      expect(imgUpload.state()).toEqual({
        fileName: '',
        imageSrc: '',
        mode: 'no-file',
        rotate: 0,
        scale: 1
      });

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(null);

      expect(onBlurSpy).toHaveBeenCalledTimes(1);

      // @ts-expect-error Test mock
      expect(imgUpload.instance().inputRef.current.value).toBe('');
    });

    it('should reset to no file and trigger the file input so user can select a new image', () => {
      jest.useFakeTimers();

      setup({ value: new File([''], 'maarten.png'), cropType: 'rect' });
      imgUpload.setState({ imageSrc: 'maarten.png', mode: 'file-selected' });

      const instance = imgUpload.instance();
      // @ts-expect-error Test mock
      instance.inputRef = { current: { click: jest.fn() } };

      jest
        // @ts-expect-error Test mock
        .spyOn(instance, 'resetFileInput')
        .mockImplementation(() => undefined);

      // @ts-expect-error Test mock
      imgUpload.find('Button').at(0).prop('onClick')();

      // @ts-expect-error Test mock
      expect(instance.resetFileInput).toHaveBeenCalledTimes(1);

      // @ts-expect-error Test mock
      expect(instance.inputRef.current.click).toHaveBeenCalledTimes(0);

      jest.runAllTimers();

      // @ts-expect-error Test mock
      expect(instance.inputRef.current.click).toHaveBeenCalledTimes(1);
    });
  });
});

test('requireImage', () => {
  const validator = requireImage('profile picture');

  // @ts-expect-error Even though it accepts Value it will be given undefined and null
  expect(validator(undefined, {})).toEqual({
    data: { label: 'profile picture' },
    key: 'ImageUpload.REQUIRED',
    fallback: 'profile picture is required'
  });
  // @ts-expect-error Even though it accepts Value it will be given undefined and null
  expect(validator(null, {})).toEqual({
    data: { label: 'profile picture' },
    key: 'ImageUpload.REQUIRED',
    fallback: 'profile picture is required'
  });

  expect(validator(new File([''], 'henkie.png'), {})).toBe(undefined);
});

test('limitImageSize', () => {
  const validator = limitImageSize(1, 'profile picture');

  // @ts-expect-error Even though it accepts Value it will be given undefined and null
  expect(validator(undefined, {})).toBe(undefined);
  // @ts-expect-error Even though it accepts Value it will be given undefined and null
  expect(validator(null, {})).toBe(undefined);

  const smallFile = new File([''], 'small.png');
  expect(validator(smallFile, {})).toBe(undefined);

  const largeFile = new File(
    ['very large image string'.repeat(100000)],
    'large.png'
  );
  expect(validator(largeFile, {})).toEqual({
    data: { fileSize: '2.2', label: 'profile picture', size: '1.0' },
    fallback:
      'profile picture image is to large. Max size is 1.0 MB image size is 2.2 MB',
    key: 'ImageUpload.SIZE_TOO_LARGE'
  });
});
