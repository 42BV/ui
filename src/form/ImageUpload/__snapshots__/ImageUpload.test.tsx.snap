// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ImageUpload ui edit as circle: Component: ImageUpload => ui => edit as circle 1`] = `
<div>
  <FormGroup
    className="img-upload"
    color="success"
    tag="div"
  >
    <Label
      for="image-uploader"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Profile photo
    </Label>
    <div
      className="d-flex justify-content-center"
    >
      <div
        onWheel={[Function]}
      >
        <a
          border={25}
          borderRadius={9000000}
          color={
            Array [
              0,
              0,
              0,
              0.5,
            ]
          }
          disableBoundaryChecks={false}
          disableCanvasRotation={true}
          disableHiDPIScaling={false}
          height={250}
          image="maarten.png"
          onImageChange={[Function]}
          onImageReady={[Function]}
          onLoadFailure={[Function]}
          onLoadSuccess={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          onPositionChange={[Function]}
          rotate={0}
          scale={1}
          width={250}
        />
      </div>
    </div>
    <FormGroup
      className="d-flex justify-content-center mt-1"
      tag="div"
    >
      <Button
        className="mt-2"
        color="secondary"
        icon="rotate_left"
        onClick={[Function]}
      />
      <Button
        className="ml-1 mt-2"
        color="secondary"
        icon="rotate_right"
        onClick={[Function]}
      />
      <Button
        className="ml-3"
        color="secondary"
        icon="cancel"
        onClick={[Function]}
      >
        Cancel
      </Button>
      <Button
        className="ml-1"
        color="primary"
        icon="done"
        onClick={[Function]}
      >
        Done
      </Button>
    </FormGroup>
    Some error
  </FormGroup>
</div>
`;

exports[`Component: ImageUpload ui edit as rect: Component: ImageUpload => ui => edit as rect 1`] = `
<div>
  <FormGroup
    className="img-upload"
    color="success"
    tag="div"
  >
    <Label
      for="image-uploader"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Profile photo
    </Label>
    <div
      className="d-flex justify-content-center"
    >
      <div
        onWheel={[Function]}
      >
        <a
          border={25}
          borderRadius={0}
          color={
            Array [
              0,
              0,
              0,
              0.5,
            ]
          }
          disableBoundaryChecks={false}
          disableCanvasRotation={true}
          disableHiDPIScaling={false}
          height={500}
          image="maarten.png"
          onImageChange={[Function]}
          onImageReady={[Function]}
          onLoadFailure={[Function]}
          onLoadSuccess={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          onPositionChange={[Function]}
          rotate={0}
          scale={1}
          width={500}
        />
      </div>
    </div>
    <FormGroup
      className="d-flex justify-content-center mt-1"
      tag="div"
    >
      <Button
        className="mt-2"
        color="secondary"
        icon="rotate_left"
        onClick={[Function]}
      />
      <Button
        className="ml-1 mt-2"
        color="secondary"
        icon="rotate_right"
        onClick={[Function]}
      />
      <Button
        className="ml-3"
        color="secondary"
        icon="cancel"
        onClick={[Function]}
      >
        CANCEL
      </Button>
      <Button
        className="ml-1"
        color="primary"
        icon="done"
        onClick={[Function]}
      >
        DONE
      </Button>
    </FormGroup>
    Some error
  </FormGroup>
</div>
`;

exports[`Component: ImageUpload ui file-selected as circle: Component: ImageUpload => ui => file-selected as circle 1`] = `
<div>
  <FormGroup
    className="img-upload"
    color="success"
    tag="div"
  >
    <Label
      for="image-uploader"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Profile photo
    </Label>
    <div
      className="d-flex justify-content-center"
    >
      <img
        alt="Profile photo"
        className="img-fluid elevated-3"
        src="maarten.png"
        style={
          Object {
            "borderRadius": "50%",
          }
        }
      />
    </div>
    <FormGroup
      className="text-center mt-1"
      tag="div"
    >
      <Button
        color="primary"
        icon="camera_roll"
        onClick={[Function]}
      >
        CHANGE
      </Button>
      <Button
        className="ml-1"
        color="danger"
        icon="delete"
        onClick={[Function]}
      >
        REMOVE
      </Button>
    </FormGroup>
    Some error
  </FormGroup>
</div>
`;

exports[`Component: ImageUpload ui file-selected as rect without label: Component: ImageUpload => ui => file-selected as rect without label 1`] = `
<div>
  <FormGroup
    className="img-upload"
    color="success"
    tag="div"
  >
    <div
      className="d-flex justify-content-center"
    >
      <img
        alt=""
        className="img-fluid elevated-3"
        src="maarten.png"
        style={
          Object {
            "borderRadius": 0,
          }
        }
      />
    </div>
    <FormGroup
      className="text-center mt-1"
      tag="div"
    >
      <Button
        color="primary"
        icon="camera_roll"
        onClick={[Function]}
      >
        Change
      </Button>
      <Button
        className="ml-1"
        color="danger"
        icon="delete"
        onClick={[Function]}
      >
        Remove
      </Button>
    </FormGroup>
    Some error
  </FormGroup>
</div>
`;

exports[`Component: ImageUpload ui file-selected as rect: Component: ImageUpload => ui => file-selected as rect 1`] = `
<div>
  <FormGroup
    className="img-upload"
    color="success"
    tag="div"
  >
    <Label
      for="image-uploader"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Profile photo
    </Label>
    <div
      className="d-flex justify-content-center"
    >
      <img
        alt="Profile photo"
        className="img-fluid elevated-3"
        src="maarten.png"
        style={
          Object {
            "borderRadius": 0,
          }
        }
      />
    </div>
    <FormGroup
      className="text-center mt-1"
      tag="div"
    >
      <Button
        color="primary"
        icon="camera_roll"
        onClick={[Function]}
      >
        Change
      </Button>
      <Button
        className="ml-1"
        color="danger"
        icon="delete"
        onClick={[Function]}
      >
        Remove
      </Button>
    </FormGroup>
    Some error
  </FormGroup>
</div>
`;

exports[`Component: ImageUpload ui no-file: Component: ImageUpload => ui => no-file 1`] = `
<div>
  <FormGroup
    className="img-upload"
    color="success"
    tag="div"
  >
    <Label
      for="image-uploader"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      Profile photo
    </Label>
    <input
      accept="image/*"
      id="image-uploader"
      onChange={[Function]}
      type="file"
    />
    <div
      className="img-upload-wrapper bg-faded text-muted"
    >
      <Icon
        icon="add_a_photo"
      />
    </div>
    Some error
  </FormGroup>
</div>
`;

exports[`Component: ImageUpload ui with empty label: Component: ImageUpload => ui => with empty label 1`] = `
<div>
  <FormGroup
    className="img-upload"
    color="success"
    tag="div"
  >
    <input
      accept="image/*"
      id="image-uploader"
      onChange={[Function]}
      type="file"
    />
    <div
      className="img-upload-wrapper bg-faded text-muted"
    >
      <Icon
        icon="add_a_photo"
      />
    </div>
    Some error
  </FormGroup>
</div>
`;

exports[`Component: ImageUpload ui without label: Component: ImageUpload => ui => without label 1`] = `
<div>
  <FormGroup
    className="img-upload"
    color="success"
    tag="div"
  >
    <input
      accept="image/*"
      id="5"
      onChange={[Function]}
      type="file"
    />
    <div
      className="img-upload-wrapper bg-faded text-muted"
    >
      <Icon
        icon="add_a_photo"
      />
    </div>
    Some error
  </FormGroup>
</div>
`;
